gtda/images/filtrations.py

```diff
@@ -1036,3 +1028,227 @@ def plot(Xt, sample=0, colorscale='greys', origin='upper',
+        'radius': {'type': Real, 'in': Interval(0, np.inf, closed='right')},
+        'metric': {'type': (str, FunctionType)},
+        'metric_params': {'type': dict},
+    }
+
+    def __init__(self, radius=3, metric='euclidean', metric_params={},
+                 n_jobs=None):
+        self.radius = radius
+        self.metric = metric
+        self.metric_params = metric_params
+        self.n_jobs = n_jobs
+
+    def _calculate_density(self, X):
+        Xd = np.zeros(X.shape)
+
+        for i, j, k in self._iterator:
+            Xd += np.roll(np.roll(
+                np.roll(X, k, axis=3), j, axis=2), i, axis=1) \
+                * self.mask_[self._range + i, self._range + j,
+                             self._range + k]
+        return Xd
+
+    def fit(self, X, y=None):
+        """Calculate :attr:`mask_` from a collection of binary images. Then,
+        return the estimator.
+
+        This method is here to implement the usual scikit-learn API and hence
+        work in pipelines.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_pixels_x, n_pixels_y [, n_pixels_z])
+            Input data. Each entry along axis 0 is interpreted as a 2D or 3D
+            binary image.
+
+        y : None
+            There is no need of a target in a transformer, yet the pipeline API
+            requires this parameter.
+
+        Returns
+        -------
+        self : object
+
+        """
+        X = check_array(X, allow_nd=True)
+        self._n_dimensions = X.ndim - 1
+        if (self._n_dimensions < 2) or (self._n_dimensions > 3):
+            warn(f"Input of `fit` contains arrays of dimension "
+                 f"{self._n_dimensions}.")
+        validate_params(
+            self.get_params(), self._hyperparameters, exclude=['n_jobs'])
+
+        self._range = int(np.ceil(self.radius))
+
+        iterator_range_list = [range(-self._range, self._range + 1)] * self._n_dimensions \
+                               for _ in range(self._n_dimensions)] \
+            + [[0] * (3 - self._n_dimensions)]
+        self._iterator = tuple(itertools.product(*iterator_range_list))
+
+        # The mask is always 3D but not the iterator.
+        self.mask_ = np.ones(tuple([2 * self._range + 1] * 3,
+                             dtype=np.bool)
+        mesh_range_list = [np.arange(0, 2 * self._range + 1)] * 3
+        self.mesh_ = np.stack(
+            np.meshgrid(*mesh_range_list), axis=3).reshape((-1, 3))
+
+        center = self._range * np.ones((1, 3))
+        self.mask_ = pairwise_distances(
+            center, self.mesh_, metric=self.metric,
+            n_jobs=1, **self.metric_params).reshape(self.mask_.shape)
+
+        self.mask_ = self.mask_ <= self.radius
+
+        padding = np.asarray([self._range] * self._n_dimensions + [0] * (3 - self._n_dimensions))
+            [*[self._range for _ in range(self._n_dimensions)],
+             *[0 for _ in range(3 - self._n_dimensions)]])
```

Comment:
There seems to be forgotten leftover code here (maybe I was unclear in my review)

Reply:
Not, you were not unclear, but you used a feature to change the code as suggestions and GitHub suggested me to commit them directly. That's what I did, but your suggestion snippets included only addition not change. xD

Reply:
Aha! Thanks for pointing that out, I'll pay more attention to this (I suspect I failed to select multiple lines).